README.md
# feathers-authentication-ldap

[![Build Status](https://travis-ci.org/feathersjs/feathers-authentication-ldap.png?branch=master)](https://travis-ci.org/feathersjs/feathers-authentication-ldap)
[![Code Climate](https://codeclimate.com/github/feathersjs/feathers-authentication-ldap/badges/gpa.svg)](https://codeclimate.com/github/feathersjs/feathers-authentication-ldap)
[![Test Coverage](https://codeclimate.com/github/feathersjs/feathers-authentication-ldap/badges/coverage.svg)](https://codeclimate.com/github/feathersjs/feathers-authentication-ldap/coverage)
[![Dependency Status](https://img.shields.io/david/feathersjs/feathers-authentication-ldap.svg?style=flat-square)](https://david-dm.org/feathersjs/feathers-authentication-ldap)
[![Download Status](https://img.shields.io/npm/dm/feathers-authentication-ldap.svg?style=flat-square)](https://www.npmjs.com/package/feathers-authentication-ldap)

> LDAP authentication strategy for feathers-authentication using Passport


## Installation

```
npm install feathers-authentication-ldap --save
```

## Documentation

### Usage

In most cases initializing the `feathers-authentication-ldap` module is as simple as doing this:

```js
app.configure(authentication(settings));
app.configure(ldap());
```

This will pull from your global `auth` object in your config file.
It will also mix in the following defaults, which should be customized.

#### Default Options

```js
{
  name: 'ldap',
  server: {
    url: 'ldap://localhost:389',
    bindDn: 'cn=anonymous',
    bindCredentials: '', // bindpw
    searchBase: 'dc=de',
    searchFilter: '(uid={{username}})',
    searchAttributes: null // optional array of props to fetch from ldap
  },
  passReqToCallback: true
}
```

### LDAP Verifier class

The `Verifier` class has a `verify` function that is the passport verify callback. In this module it gets called after LDAP authentication succeeds. By default it does nothing but you can overwrite it to make furthers validation
checks. See [examples/app.js](examples/app.js#L56). 

### Usage with `feathers-authentication-jwt`

To authenticate following requests using the jwt use `feathers-authentication-jwt`.  This plugin depends on the `users` Service to populate the user entity.

To get rid of this dependency and store necessary data in the JWT payload see
[examples/app.js](examples/app.js#L47) and [examples/app.js](examples/app.js#L79).
 

## Simple Example

Here's an example of a Feathers server that uses `feathers-authentication-ldap`. 

```js
const feathers = require('feathers');
const rest = require('feathers-rest');
const hooks = require('feathers-hooks');
const bodyParser = require('body-parser');
const errorHandler = require('feathers-errors/handler');
const errors = require('feathers-errors');
const auth = require('feathers-authentication');
const ldap = require('feathers-authentication-ldap');

// Initialize the application
const app = feathers();

// Load configuration, usually done by feathers-configuration
app.set('auth', {
  secret: "super secret",
  ldap: {
    server: {
      url: 'ldap://localhost:389',
      bindDn: 'cn=anonymous',
      bindCredentials: '', // bindpw
      searchBase: 'dc=de',
      searchFilter: '(uid={{username}})'
    }
  }
});

app.configure(rest())
  .configure(hooks())
  .use(bodyParser.json())
  .use(bodyParser.urlencoded({ extended: true }))

  // Configure feathers-authentication with ldap
  .configure(auth(app.get('auth')))
  .configure(ldap())

  .use(errorHandler());

// Authenticate the user using the LDAP strategy
// and if successful return a JWT.
app.service('authentication').hooks({
  before: {
    create: [
      auth.hooks.authenticate('ldap')
    ]
  }
});

app.listen(3030);
console.log('Feathers authentication with LDAP auth started on 127.0.0.1:3030');
```

```shell
# Request a token
curl -H "Content-Type: application/json" \
     -X POST \
     -d '{"username":"ldap-user@example.com","password":"admin"}' \
     http://localhost:3030/authentication
```

## License

Copyright (c) 2016

Licensed under the [MIT license](LICENSE).


CHANGELOG.md
# Change Log

## [v0.2.3](https://github.com/feathersjs/feathers-authentication-ldap/tree/v0.2.3) (2017-03-28)
[Full Changelog](https://github.com/feathersjs/feathers-authentication-ldap/compare/v0.2.2...v0.2.3)

## [v0.2.2](https://github.com/feathersjs/feathers-authentication-ldap/tree/v0.2.2) (2017-03-28)
[Full Changelog](https://github.com/feathersjs/feathers-authentication-ldap/compare/v0.2.1...v0.2.2)

**Closed issues:**

- No ldap provider when used with feathers-cli generate [\#4](https://github.com/feathersjs/feathers-authentication-ldap/issues/4)

**Merged pull requests:**

- Update all dependencies 🌴 [\#7](https://github.com/feathersjs/feathers-authentication-ldap/pull/7) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Updated deps [\#6](https://github.com/feathersjs/feathers-authentication-ldap/pull/6) ([psi-4ward](https://github.com/psi-4ward))

## [v0.2.1](https://github.com/feathersjs/feathers-authentication-ldap/tree/v0.2.1) (2016-12-14)
[Full Changelog](https://github.com/feathersjs/feathers-authentication-ldap/compare/v0.2.0...v0.2.1)

## [v0.2.0](https://github.com/feathersjs/feathers-authentication-ldap/tree/v0.2.0) (2016-12-14)
[Full Changelog](https://github.com/feathersjs/feathers-authentication-ldap/compare/v0.1.0...v0.2.0)

## [v0.1.0](https://github.com/feathersjs/feathers-authentication-ldap/tree/v0.1.0) (2016-12-14)
**Merged pull requests:**

- Consistency touchups [\#3](https://github.com/feathersjs/feathers-authentication-ldap/pull/3) ([ekryski](https://github.com/ekryski))
- init [\#1](https://github.com/feathersjs/feathers-authentication-ldap/pull/1) ([psi-4ward](https://github.com/psi-4ward))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

test/index.test.js
import chai, {expect} from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import ldapAuth from '../src';
import feathers from 'feathers';
import authentication from 'feathers-authentication';
import passportLdap from 'passport-ldapauth';
chai.use(sinonChai);

describe('feathers-authentication-ldap', () => {
  it('is CommonJS compatible', () => {
    expect(typeof require('../lib')).to.equal('function');
  });

  it('basic functionality', () => {
    expect(typeof ldapAuth).to.equal('function');
    expect(typeof ldapAuth()).to.equal('function');
  });

  it('exposes the default Verifier fn', () => {
    expect(typeof ldapAuth.Verifier).to.equal('function');
  });

  it('exposes the defaults obj', () => {
    expect(typeof ldapAuth.defaults).to.equal('object');
  });

  describe('defaultVerifier fn', () => {
    it('should cb without error', function (done) {
      let verifier = new ldapAuth.Verifier();
      verifier.verify({body: {username: 'test'}}, {name: 'test'}, done);
    });
  });

  describe('jwtVerifier fn', () => {
    it('should cb without error', function (done) {
      let jwtVerifier = new ldapAuth.JWTVerifier();
      jwtVerifier.verify({body: {username: 'test'}}, {name: 'test'}, done);
    });
  });

  describe('initialization', () => {
    let app;

    beforeEach(() => {
      app = feathers();
      app.configure(authentication({secret: 'supersecret'}));
    });

    it('throws an error if passport has not been registered', () => {
      expect(() => {
        feathers().configure(ldapAuth());
      }).to.throw();
    });

    it('registers the ldap passport strategy', () => {
      sinon.spy(app.passport, 'use');
      sinon.spy(passportLdap, 'Strategy');
      app.configure(ldapAuth());
      app.setup();

      expect(passportLdap.Strategy).to.have.been.calledOnce;
      expect(app.passport.use).to.have.been.calledWith('ldap');

      app.passport.use.restore();
      passportLdap.Strategy.restore();
    });

    it('registers the strategy options', () => {
      sinon.spy(app.passport, 'options');
      app.configure(ldapAuth());
      app.setup();

      expect(app.passport.options).to.have.been.calledOnce;

      app.passport.options.restore();
    });
  });
});


src/index.js
import Debug from 'debug';
import merge from 'lodash.merge';
import omit from 'lodash.omit';
import DefaultVerifier from './verifier';
import { Strategy as LdapStrategy } from 'passport-ldapauth';

const debug = Debug('feathers-authentication-ldap');
const defaults = {
  name: 'ldap',
  server: {
    url: 'ldap://localhost:389',
    bindDn: 'cn=anonymous',
    bindCredentials: '',
    searchBase: 'dc=de',
    searchFilter: '(uid={{username}})',
    searchAttributes: null
  },
  passReqToCallback: true
};

class JWTVerifier {
  // JWT Verifier replacement
  // to disable population from user service
  verify (req, payload, done) {
    done(null, { payload });
  }
}

// Export ldap-auth init function
export default function init (options = {}) {
  return function ldapAuth () {
    const app = this;
    const _super = app.setup;

    if (!app.passport) {
      throw new Error(`Can not find app.passport. Did you initialize feathers-authentication before feathers-authentication-ldap?`);
    }

    // Construct ldapSettings for passport ldap strategy
    let name = (typeof(options) !== 'function' && options.name) || defaults.name;
    let authOptions = app.get('auth') || {};
    let ldapOptions = authOptions[name] || {};
    const ldapSettings = merge({}, defaults, ldapOptions, (typeof options === 'function' ? {} : omit(options, ['Verifier'])));
    const Verifier = options.Verifier || DefaultVerifier;
    const asyncOptions = function (req, callback) {
      options(req)
      .then(function (opts) {
        callback(null, merge({}, ldapSettings, opts));
      })
      .catch(function (err) {
        callback(err, ldapSettings);
      });
    };
    const ldapStrategySettings = (typeof options === 'function' ? asyncOptions : ldapSettings);

    // plugin setup: register strategy in feathers passport
    app.setup = function () {
      // be sure feathers setup was called
      let result = _super.apply(this, arguments);
      let verifier = new Verifier(app, ldapSettings);

      if (!verifier.verify) {
        throw new Error(`Your verifier must implement a 'verify' function. It should have the same signature as function(request, user, done)`);
      }

      // Register 'ldap' strategy with passport
      debug('Registering ldap authentication strategy with options:', ldapSettings);
      app.passport.use(ldapSettings.name, new LdapStrategy(ldapStrategySettings, verifier.verify.bind(verifier)));
      app.passport.options(ldapSettings.name, ldapSettings); // do we need this ??

      return result;
    };
  };
}

// Exposed Modules
Object.assign(init, {
  defaults,
  Verifier: DefaultVerifier,
  JWTVerifier
});


example/app.js
const feathers = require('feathers');
const rest = require('feathers-rest');
const hooks = require('feathers-hooks');
const bodyParser = require('body-parser');
const errorHandler = require('feathers-errors/handler');
const errors = require('feathers-errors');
const auth = require('feathers-authentication');
const jwt = require('feathers-authentication-jwt');
const ldap = require('../lib/index');

// Initialize the application
const app = feathers();

// Load configuration, usually done by feathers-configuration
app.set('auth', {
  secret: "super secret",
  ldap: {
    server: {
      url: 'ldap://localhost:389',
      bindDn: 'cn=anonymous',
      bindCredentials: '', // bindpw
      searchBase: 'dc=de',
      searchFilter: '(uid={{username}})',
      searchAttributes: null // set an array of props to fetch from ldap
    }
  }
});

class CustomVerifier {
  verify(req, user, done) {
    // ldap auth was successful
    console.log('LDAP User:', user);

    // add custom verification logic
    if(true) {
      return done(null, user, {username: req.body.username});
    } else {
      const err = new errors.Forbidden('Youre are not allowed');
      return done(err);
    }
  }
}

app.configure(rest())
  .configure(hooks())
  // Needed for parsing bodies (login)
  .use(bodyParser.json())
  .use(bodyParser.urlencoded({ extended: true }))

  // Configure feathers-authentication
  .configure(auth(app.get('auth')))

  // Enable JWT auth
  .configure(jwt({
    // Optional: Overwrite the JWT Verifier to disable
    // population from user service
    Verifier: ldap.JWTVerifier
  }))

  // Enable LDAP auth
  .configure(ldap({
    // Optional: overwrite Verifier function
    Verifier: CustomVerifier
  }))

  .use(errorHandler());


// Authenticate the user using the LDAP strategy
// and if successful return a JWT.
app.service('authentication').hooks({
  before: {
    create: [
      auth.hooks.authenticate('ldap'),
      function(hook) {
        // store some user data in JWT payload
        hook.data.payload = {
          uid: hook.params.user.uid,
          name: hook.params.user.cn,
          mail: hook.params.user.mail
        };
        return Promise.resolve(hook);
      }
    ]
  }
});

app.listen(3030);
console.log('Feathers authentication with LDAP auth started on 127.0.0.1:3030');
